types/single-spa-react.d.ts

```diff
@@ -1,21 +1,21 @@
 import * as React from "react";
-import * as ReactDOM from "react-dom";
+import * as ReactDOMClient from "react-dom/client";
 import { AppProps, CustomProps, LifeCycleFn } from "single-spa";
 
 export const SingleSpaContext: React.Context<CustomProps & AppProps>;
 
+export interface ReactAppOrParcel<ExtraProps> {
+  bootstrap: LifeCycleFn<ExtraProps>;
+  mount: LifeCycleFn<ExtraProps>;
+  unmount: LifeCycleFn<ExtraProps>;
+  update?: LifeCycleFn<ExtraProps>;
+}
+
 export interface SingleSpaReactOpts<RootComponentProps> {
   React: typeof React;
-  ReactDOM: typeof ReactDOM;
-  rootComponent?:
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>;
-  loadRootComponent?: (
-    props: RootComponentProps
-  ) => Promise<
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>
-  >;
+  ReactDOMClient: typeof ReactDOMClient;
```

Comment:
As far as I know, typescript doesn't have a way for us to provide the correct types for these two properties. The single-spa-react project supports React 15-18, which has a variety of types. In an ideal world, we'd do something like this:

```ts
import * as React15 from 'react15';
import * as React16 from 'react16';
import * as React17 from 'react17';
import * as React18 from 'react18';
import * as ReactDOM15 from 'react15';
import * as ReactDOM16 from 'react16';
import * as ReactDOM17 from 'react17';
import * as ReactDOM18 from 'react18';

export interface SingleSpaReactOpts {
  React: typeof React15 | typeof React16 | typeof React17 | typeof React18;
  ReactDOM: typeof ReactDOM15 | typeof ReactDOM16 | typeof ReactDOM17 | typeof ReactDOM18;
}
```

However, I don't think it's possible to install all the different versions of `@types/react` and then get typescript to properly alias them in a generally reusable way.

For this reason, i propose just switching to `any` for the `React` and `ReactDOM` types

Reply:
One alternative to using `any` might be to use like types for `ReactDOM`. Maybe something along these lines?

```typescript
interface ReactDOMLike {
  render: typeof ReactDOM.render;
}

interface ReactDOMClientLike {
  createRoot: typeof ReactDOMClient.createRoot;
  hydrate: typeof ReactDOMClient.hydrateRoot;
}

export interface SingleSpaReactOpts {
  ReactDOM: ReactDOMLike | ReactDOMClientLike;
}
```

Reply:
@limulus how would that work if I'm using React 17 still and my version of react types doesn't have `ReactDOMClient`? The whole point is that we want backwards compatibility so using ReactDOMClient type is not possible without making single-spa-react incompatible with react 17 and below. At least that's my understanding.

Reply:
@filoxo Yeah, you are right, as I wrote it that would not work. I was going to suggest that you could inline the types of the `createRoot`, `hydrate`, and `render` functions so that you would not need to pull those types from the React types package. But I took a look at your solution from the other PR and that seems reasonable enough to me!